### ISSUE TITLE
Layer normalization: compute mean and standard deviation in single pass

### ISSUE THREAD

**Issue by kpu** [marian-nmt/marian-dev#429]
Had a /fun/ discussion with @bzhangGo over layer normalization in the group meeting.  

Layer normalization currently computes mean then standard deviation which is a two-pass approach:

https://github.com/marian-nmt/marian-dev/blob/63e1cfe4adaf11cf6006f6325c4e60a56bd79e96/src/tensors/gpu/tensor_operators.cu#L1497

Mean and standard deviation can be computed in one pass by computing sum and sum of squares, reducing memory access. @rsennrich pointed out that this classic algorithm isn't that numerically stable (since the squares get big) but there are fancier single-pass algorithms: http://dl.acm.org/citation.cfm?id=3221269.3223036 .  

Propose @XapaJIaMnu and @bzhangGo talk about this, but I can't assign @bzhangGo because he's not on the repo.

**Comment by emjotde** [comment]
I can send an invite. LN is barely manifesting during profiling though. Unless I am have been doing that wrong.

**Comment by emjotde** [comment]
The numeric stability might good to have though.

**Comment by emjotde** [comment]
Sent an invite to @bzhangGo

**Comment by rsennrich** [comment]
LN overhead isn't that substantial for Transformer, but about 10% for a deep RNN with Marian.

I'm not advocating for doing a one-pass standard deviation though (which would be either unstable or more complex to implement) - Biao has shown that you get the same quality if you simply ignore the mean, and just take the root mean square: https://openreview.net/forum?id=SygkZ3MTJE

**Comment by XapaJIaMnu** [comment]
It looks like, it's just two CUDA kernels that do all of that. If @bzhangGo wants some help implementing it, I can sit down with him next week for a few hours and help.

**Comment by emjotde** [comment]
Ah, so it's the one pass version that is numerically unstable. Got it. 

The 10% is inference or training? I recently fixed a couple of things getting rid of atomic adds in the LN gradient which should make things a bit better.

**Comment by emjotde** [comment]
From the abstract of that paper, "2.5%∼15% on different models against LayerNorm". This is total runtime or LayerNorm runtime? Are we talking here about improving runtime by 0.25 to 1.5% total if LN takes 10% in Marian?

**Comment by emjotde** [comment]
Total runtime. OK, Worth a try. Let me know if anyone needs help.

**Comment by hieuhoang** [comment]
@rsennrich quickly looking at equ 3 in the paper, can't you calc the variance in 1 pass using old skool var = 1/n [sum( x^2) -2u sum(x) + u]
?

**Comment by rsennrich** [comment]
that's the texbook algorithm in the reference above (Schubert and Gertz), and yes, it's numerically unstable.

**Comment by hieuhoang** [comment]
Ah yes. Exploding sum squared

**Comment by emjotde** [comment]
I had some stability issues with LayerNorm recently, especially with the gradients. I would be interested in reviving this thread. Anything interesting to look at?

**Comment by hieuhoang** [comment]
it smells similar to the 1 pass softmax i did in amun.

**Comment by emjotde** [comment]
Yeah, I want it for stability-reasons rather than speed. Assuming it helps with gradients too.

**Comment by emjotde** [comment]
Ah, read the thread again. Replacing this will not help with stability.

**Comment by emjotde** [comment]
Allowing myself to close this, as it seems there is not much performance nor stability to be gained. Feel free to re-open if you disagree.
